### ISSUE TITLE
[WIP] [GH-38] Implement undefined variable rule

### ISSUE THREAD

**Issue by manuelpichler** [phpmd/phpmd#65]
There is a class of bugs when you use a variable that is undefined. Some PHP functions allow this when the parameter is passed by reference, however it is still decouraged not define this variables beforehand for clarity.

This rule implements checking for undefined variables. This is done by first collecting all variable assignments, then all parameters of the method/function.\n\nIn the case of the following example, PHP_PMD will trigger two errors, unused local variable and undefined variable:

``` php
function foo()
    {
        echo $bar;
    }
```

therefore propose to work on this patch further by modifying the UnusedLocalVariable to only warn about variable assignments that are not used and keep the undefined variable rule to catch the error class shown above.\n\nPossible Improvements:\n\n\* Check if the variable assignment is actually BEFORE the variable usage. Currently it only checks that there is an assignment, it could be after the usage though.

Synced from GitHub: https://github.com/phpmd/phpmd/issues/64

Synced from PivotalTracker: http://www.pivotaltracker.com/story/show/42432495

**Comment by ava-dylang** [comment]
:+1:

**Comment by shufenghua** [comment]
also need it

**Comment by oligriffiths** [comment]
Hey, what's the state of play with this? I notice it keeps getting milestone bumped, seems like a very useful rule.

**Comment by ravage84** [comment]
@oligriffiths you are welcome to contribute a PR.

**Comment by ravage84** [comment]
Moving to Future for now. If we are going to implement this, we can have a look at the proposed changes in https://github.com/phpmd/phpmd/pull/64/files.

**Comment by liaol** [comment]
need it
